src/shared/ui/pickerPrompter.ts

```diff
@@ -0,0 +1,255 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+
+import { WIZARD_BACK } from '../wizards/wizard'
+import { QuickInputButton, PrompterButtons } from './buttons'
+import { Prompter, PromptResult } from './prompter'
+import { applyPrimitives } from '../utilities/collectionUtils'
+
+/** Additional options to configure the `QuickPick` beyond the standard API */
+interface AdditionalQuickPickOptions<T = never> {
+    title?: string
+    value?: string
+    step?: number
+    placeholder?: string
+    totalSteps?: number
+    buttons?: PrompterButtons<T>
+    customInputSettings?: {
+        label: string
+        transform: (v: string) => PromptResult<T>
+    }
+}
+
+export type ExtendedQuickPickOptions<T> = Omit<vscode.QuickPickOptions, 'buttons' | 'canPickMany' | 'placeHolder'> &
+    AdditionalQuickPickOptions<T>
```

Comment:
> why is placeholder redefined on AdditionalQuickPickOptions

The options have `placeHolder`, but the property name on the QuickPick is `placeholder`. Since we're doing assignment directly from options the names need to be correct.

> don't see buttons defined on vscode.QuickPickOptions, but it's omitted?

Remnant of earlier revisions where the options were based off the QuickPick interface directly. Removed.

> why omit canPickMany? is it a TODO/future?

Yup, have not implemented the `MultiQuickPick` prompter since we do not use it.

Since `Prompter` is a high-level abstraction, the options should be called `QuickPickPrompterOptions` since they don't necessarily apply to the `QuickInput` prompter.

Reply:
The indirection is also really undesirable. I'm thinking I'll just define the interface directly instead of extending the native one. This is a case where code (type) re-use is not a good thing.

Reply:
Extending is good IMO, it documents that this is roughly parallel to QuickPickOptions. But just the AdditionalQuickPickOptions named type seems unnecessary.